Python/wolfssl-python-3.8.5.patch

```diff
@@ -0,0 +1,1995 @@
+                 s.connect((HOST, server.port))
+-        self.assertGreaterEqual(self.keylog_lines(), 21)
++        if ssl.IS_WOLFSSL:
++            # Does not include EXPORTER
++            self.assertEqual(self.keylog_lines(), 17)
++        else:
++            self.assertGreaterEqual(self.keylog_lines(), 21)
+ 
+         client_context.keylog_filename = None
+         server_context.keylog_filename = None
+diff -Naur a/Modules/_hashopenssl.c b/Modules/_hashopenssl.c
+--- a/Modules/_hashopenssl.c	2021-09-16 08:35:22.346325831 -0700
++++ b/Modules/_hashopenssl.c	2021-09-10 14:24:36.252210948 -0700
+@@ -18,6 +18,10 @@
+ #include "hashlib.h"
+ #include "pystrhex.h"
+ 
++#ifdef HAVE_WOLFSSL
++#include "wolfssl/options.h"
++#include "openssl/opensslconf.h"  /* for OPENSSL_THREADS define */
++#endif
+ 
+ /* EVP is the preferred interface to hashing in OpenSSL */
+ #include <openssl/evp.h>
+@@ -30,7 +34,8 @@
+ #  error "OPENSSL_THREADS is not defined, Python requires thread-safe OpenSSL"
+ #endif
+ 
+-#if (OPENSSL_VERSION_NUMBER < 0x10100000L) || defined(LIBRESSL_VERSION_NUMBER)
++#if ((OPENSSL_VERSION_NUMBER < 0x10100000L) || defined(LIBRESSL_VERSION_NUMBER)) && \
++    !defined(HAVE_WOLFSSL)
+ /* OpenSSL < 1.1.0 */
+ #define EVP_MD_CTX_new EVP_MD_CTX_create
+ #define EVP_MD_CTX_free EVP_MD_CTX_destroy
+@@ -38,7 +43,8 @@
+ 
+ #define MUNCH_SIZE INT_MAX
+ 
+-#ifdef NID_sha3_224
++#if defined(NID_sha3_224) || \
++    (defined(HAVE_WOLFSSL) && defined(WOLFSSL_SHA3))
+ #define PY_OPENSSL_HAS_SHA3 1
+ #endif
+ 
+@@ -126,7 +132,8 @@
+     case NID_sha512:
+         name ="sha512";
+         break;
+-#ifdef NID_sha512_224
++#if defined(NID_sha512_224) || (defined(HAVE_WOLFSSL) && \
++    defined(WOLFSSL_SHA512) && !defined(WOLFSSL_NOSHA512_224))
+     case NID_sha512_224:
+         name ="sha512_224";
+         break;
+diff -Naur a/Modules/_ssl.c b/Modules/_ssl.c
+--- a/Modules/_ssl.c	2021-09-16 08:35:22.342325987 -0700
++++ b/Modules/_ssl.c	2021-09-16 08:32:37.520740247 -0700
+@@ -54,6 +54,10 @@
+ #endif
+ 
+ /* Include OpenSSL header files */
++#ifdef HAVE_WOLFSSL
++#include "wolfssl/options.h"
++#include "openssl/opensslconf.h"  /* for OPENSSL_THREADS define */
++#endif
+ #include "openssl/rsa.h"
+ #include "openssl/crypto.h"
+ #include "openssl/x509.h"
+@@ -170,17 +174,42 @@
+ /* SNI support (client- and server-side) appeared in OpenSSL 1.0.0 and 0.9.8f
+  * This includes the SSL_set_SSL_CTX() function.
+  */
+-#ifdef SSL_CTRL_SET_TLSEXT_HOSTNAME
++#if defined(SSL_CTRL_SET_TLSEXT_HOSTNAME) || \
++    (defined(HAVE_WOLFSSL) && defined(HAVE_SNI))
++#undef HAVE_SNI
+ # define HAVE_SNI 1
+ #else
+ # define HAVE_SNI 0
+ #endif
+ 
+-#ifdef TLSEXT_TYPE_application_layer_protocol_negotiation
++#if !defined(HAVE_ALPN) || defined(HAVE_WOLFSSL)
++#if defined(TLSEXT_TYPE_application_layer_protocol_negotiation) || \
++    defined(HAVE_WOLFSSL)
++#ifdef HAVE_WOLFSSL
++/* wolfSSL defines HAVE_ALPN, but without a value. Undef and redefine here */
```

Comment:
A comment like this would also be useful above for the `HAVE_SNI` logic, which is doing the same type of thing.